### DISCUSSION TITLE
DateTime parsing

### DISCUSSION THREAD

**Discussion by edfenergy-richardcannock** [quickfix-j/quickfixj#446] [Q&A]
We are attempting to use QuickFix\J to retrieve a trade capture report for gas trades for the last three days from the ICE intercontinental exchange as follows:

quickfix.field.TradeRequestID id = new quickfix.field.TradeRequestID("SubscriptionId");
		
quickfix.field.TradeRequestType reporttype =  new TradeRequestType(TradeRequestType.MATCHED_TRADES_MATCHING_CRITERIA_PROVIDED_ON_REQUEST);
quickfix.fix44.TradeCaptureReportRequest req = new quickfix.fix44.TradeCaptureReportRequest(id,reporttype); 
quickfix.field.SubscriptionRequestType subType = new quickfix.field.SubscriptionRequestType(SubscriptionRequestType.SNAPSHOT);
		
		// Add date filters.
		NoDates dates = new NoDates(2); 
		req.set(dates);
		
		quickfix.fix44.TradeCaptureReportRequest.NoDates nd1 = new quickfix.fix44.TradeCaptureReportRequest.NoDates();
		quickfix.field.TransactTime start = new quickfix.field.TransactTime(LocalDateTime.now().minusDays(3));
		//quickfix.field.TransactTime start = new quickfix.field.TransactTime(LocalDateTime.of(2021, Month.NOVEMBER, 30, 9, 30));
		
		nd1.set(start);
		
		req.addGroup(nd1);
		
		
		quickfix.fix44.TradeCaptureReportRequest.NoDates nd2 = new quickfix.fix44.TradeCaptureReportRequest.NoDates();
		quickfix.field.TransactTime end = new quickfix.field.TransactTime(LocalDateTime.now());
		nd2.set(end);
		
		req.addGroup(nd2);
		
		
		req.set(subType);
		return Session.sendToTarget(req,this.sessId);	

This results in the following error from the exchange, which seems to relate to the dates passed in:

8=FIX.4.49=18035=AQ49=ICE34=252=20211201-11:00:43.34256=1494357=EdfDev01568=SubscriptionId569=1749=99750=258=Error while parsing trade date 'null'. Format: YYYYMMDD-HH:MM:SS[.sss]10=009)

I'm not sure if this is due to the values passed into TransactTime and as a result some bug in QuickFix\J

Apologies if this is best address via the mailing list.

**Comment by the-thing** [comment]
The error states "trade date" so it is not complaining about TransactTime, but TradeDate(75) which is part of NoDates repeating group.

**Reply by edfenergy-richardcannock** [reply]
Thanks. I'm not explicitly setting any trade date in any of the NoDAtes groups but i can. TradeDate constructor seems to indicate a string. Is there any example of this? Assume it's some output of SimpleDateFormatter but is there any examples.

**Reply by the-thing** [reply]
Assuming your TradeDate is LOCALMKTDATE in FIX dictionary then the format should be "YYYYMMDD".

Also you should code against exchange FIX spec. It seems that venue requires you to populate it, but you don't. If responses from the venue do not make sense - contact exchange first.

**Reply by edfenergy-richardcannock** [reply]
Thank you. I've tried both YYYYMMDD and the format suggested by the error:

Error while parsing trade date 'null'. Format: YYYYMMDD-HH:MM:SS

to no avail. They still give the same error. 

I've gone back to the ICE exchange again for clarity on what their API actually expects....

**Reply by edfenergy-richardcannock** [reply]
I've followed the suggestions above, have sent the FIX output to the exchange, which looks like it should work. The only other possibility I can think of is that the callback signature for the returned report is wrong:

I have the following onMessage method signature for a TradeReport:

public void onMessage(quickfix.fix44.TradeCaptureReport acq,  SessionID arg1)
{
		System.out.println("Trade Report Recived please");
}

does this look ok?

**Reply by chrjohn** [reply]
> The only other possibility I can think of is that the callback signature for the returned report is wrong:

The exchange does not know anything about your application or its callback methods. So why do you think they will reject a message because of a callback method that you have?

It would probably be more helpful to see how the ICE API for that message looks like to tell if you  constructed your message the wrong way.

**Reply by edfenergy-richardcannock** [reply]
This was more about the returned data. I have confirmed with ICE now that I am sending the right data. I am now interested in making sure that the received data will be received and that the signature of the method:

public void onMessage(quickfix.fix44.TradeCaptureReport acq, SessionID arg1)
{
System.out.println("Trade Report Recived please");
}

looked correct.
